### DISCUSSION TITLE
StaticFiles always checks package directories

### DISCUSSION THREAD

**Discussion by falkben** [encode/starlette#1821] [Potential Issue] [answered]
`StaticFiles` always checks for the existence of package directories, regardless of whether `check_dir` is set to True or False.

as an example, this raises an exception:

```python
app = StaticFiles(packages=[("tests", "no_such_directory")], check_dir=False)
```

Whereas I would have expected it to be able to be create the `app` instance without errors when `check_dir` is set to `False`.

Before I create an Issue or submit a Pull Request, I just wanted to confirm whether this was something that was intended or not.

It looks like the `check_dir` flag was added [several years ago](https://github.com/encode/starlette/pull/195) but was not included when [support for packages were added](https://github.com/encode/starlette/pull/415) to `StaticFiles`

**Comment by Kludex** [comment]
May I ask on which case we would like to have `check_dir=False` for `packages`?

**Reply by falkben** [reply]
I was thinking about the following case:

You have docs or static HTML you want to mount into your application, but those directories are generated either during deployment or possibly in a later build process of your CI, or maybe have additional dependencies. These directories are also created relative to the current path of your application code, and you'd like to not worry about the full path to your static directories (thus, you want to treat them as a "packages").

Meanwhile, when testing your application, either locally or in parts of your CI where the static files are not available, you would like to be able to run your tests without needing to generate those files.

Since we have the `check_dir` flag, I figured it should be supported behavior to avoid checking these directories.

Currently, if you don't have the packages present, you cannot run your tests since `StaticFiles` will raise an exception.

**Reply by Kludex** [reply]
PR welcome. :+1:

**Reply by Kludex** [reply]
Why an empty folder with an `__init__.py` would be bad here? I'm failing to see how this use-case would be on the `packages` side - besides comfort on the path size.

**Comment by Kludex** [comment, marked as answer]
I've created https://github.com/encode/starlette/issues/2001 from this discussion. :+1:
